File: pixelpipes/image/render.cpp

#include <opencv2/imgproc.hpp>

#include <pixelpipes/image.hpp>
#include <pixelpipes/geometry.hpp>

namespace pixelpipes {

/**
 * @brief Draw a polygon to a canvas of a given size.
 * 
 */
SharedVariable Polygon(std::vector<SharedVariable> inputs, ContextHandle context) noexcept(false) {
  

    VERIFY(inputs.size() == 3, "Incorrect number of parameters");
    VERIFY(List::is_list(inputs[0], Point2DType), "Not a list of points");

    int width = Integer::get_value(inputs[1]);
    int height = Integer::get_value(inputs[2]);

    std::vector<cv::Point2f> points = List::cast(inputs[0])->elements<cv::Point2f>();

    try {

        std::vector<cv::Point> v(points.begin(), points.end());

        cv::Mat mat = cv::Mat::zeros(cv::Size(width, height), CV_8UC1);

        cv::fillPoly(mat, std::vector<std::vector<cv::Point>>({v}), cv::Scalar(255,255,255));

        return std::make_shared<Image>(mat);

    } catch (cv::Exception& cve) {
        throw VariableException(cve.what());
    }

}

REGISTER_OPERATION_FUNCTION("polygon", Polygon);


/*
NOISE GENERATION
*/

/**
 * @brief Creates a single channel image with values sampled from normal distribution.
 * 
 */
SharedVariable NormalNoise(std::vector<SharedVariable> inputs, ContextHandle context) noexcept(false) {

    VERIFY(inputs.size() == 4, "Incorrect number of parameters");

    int width = Integer::get_value(inputs[0]);
    int height = Integer::get_value(inputs[1]);
    float mean = Float::get_value(inputs[2]);
    float std = Float::get_value(inputs[3]);

	cv::RNG generator(context->random());
	cv::Mat noise(height, width, CV_64F);
	generator.fill(noise, cv::RNG::NORMAL, mean, std);

    return std::make_shared<Image>(noise);
}

REGISTER_OPERATION_FUNCTION_WITH_BASE("normal_noise", NormalNoise, StohasticOperation);

/**
 * @brief Creates a single channel image with values sampled from uniform distribution.
 * 
 */
SharedVariable UniformNoise(std::vector<SharedVariable> inputs, ContextHandle context) noexcept(false) {
    
    VERIFY(inputs.size() == 4, "Incorrect number of parameters");

    int width = Integer::get_value(inputs[0]);
    int height = Integer::get_value(inputs[1]);
    float min = Float::get_value(inputs[2]);
    float max = Float::get_value(inputs[3]);

	cv::RNG generator(context->random());
	cv::Mat noise(height, width, CV_64F);
	generator.fill(noise, cv::RNG::UNIFORM, min, max);

    return std::make_shared<Image>(noise);
}

REGISTER_OPERATION_FUNCTION_WITH_BASE("uniform_noise", UniformNoise, StohasticOperation);


SharedVariable Linear(std::vector<SharedVariable> inputs, ContextHandle context, bool flip) noexcept(false) {
    
    VERIFY(inputs.size() == 4, "Incorrect number of parameters");

    int width = Integer::get_value(inputs[0]);
    int height = Integer::get_value(inputs[1]);

    float min = Float::get_value(inputs[2]);
    float max = Float::get_value(inputs[3]);

	cv::Mat result(height, width, CV_64F);

    double *data = &result.at<double>(0, 0);

    if (flip) {
        for(int y = 0; y < result.rows; y++){
            float v = ((float) y / (float) (result.rows-1)) * (max - min) + min;
            for(int x = 0; x < result.cols; x++){  
                *data = v;
                data++;
            }
        }

    } else {

        for(int y = 0; y < result.rows; y++){
            for(int x = 0; x < result.cols; x++){  
                *data = ((float) x / (float) (result.cols-1)) * (max - min) + min;
                data++;
            }
        }
    }

    return std::make_shared<Image>(result);
}

REGISTER_OPERATION_FUNCTION("linear", Linear, bool);




}

File: pixelpipes/image/render.py

from ..graph import Input, Operation, SeedInput
from .. import types

class GaussianNoise(Operation):
    """
    Creates a single channel image with values sampled from gaussian distribution.
    """

    width = Input(types.Integer(), description="Image width")
    height = Input(types.Integer(), description="Image height")
    mean = Input(types.Float(), default=0, description="Mean of distribution")
    std = Input(types.Float(), default=1, description="Standard deviation of values")
    seed = SeedInput()
    
    def operation(self):
        return "opencv:gaussian_noise",

class UniformNoise(Operation):
    """
    Creates a single channel image with values sampled from uniform distribution.
    """

    width = Input(types.Integer(), description="Image width")
    height = Input(types.Integer(), description="Image height")
    min = Input(types.Float(), default=0, description="Minimum value")
    max = Input(types.Float(), default=1, description="Maximum value")
    seed = SeedInput()

    def operation(self):
        return "opencv:uniform_noise",

class BinaryNoise(Operation):
    """
    Creates a single channel image where a random percentage of values are set to 1.
    """

    width = Input(types.Integer(), description="Image width")
    height = Input(types.Integer(), description="Image height")
    positive = Input(types.Float(), default=0.5, description="Percentage of non-zero values, between 0 and 1")
    seed = SeedInput()

    def operation(self):
        return "opencv:binary_noise",

class LinearImage(Operation):
    """Generate an image with linearly progressing values from min to max.
    """

    width = Input(types.Integer(), description="Image width")
    height = Input(types.Integer(), description="Image height")
    min = Input(types.Float(), description="Minimum value")
    max = Input(types.Float(), description="Maximum value")
    flip = Input(types.Boolean(), default=False, description="Horizontal or vertical gradient")

    def operation(self):
        return "opencv:linear_image",

class PolygonMask(Operation):
    """Draw a polygon to a canvas of a given size
    """

    source = Input(types.Points(), description="List of points")
    width = Input(types.Integer(), description="Image width")
    height = Input(types.Integer(), description="Image height")
    thickness = Input(types.Integer(), default=0, description="Thickness of the line (0 fills the polygon)")

    def operation(self):
        return "opencv:polygon_mask",

class PointsMask(Operation):
    """Generate a list of points for a polygon
    """

    num_points = Input(types.Integer(), description="Number of points")
    width = Input(types.Integer(), description="Image width")
    height = Input(types.Integer(), description="Image height")
    size = Input(types.Integer(), default=1, description="Size of each point")

    def operation(self):
        return "opencv:points_mask",